### ISSUE TITLE
Out of memory for "changed" but not for "newer"!?

### ISSUE THREAD

**Issue by zstadler** [drolbr/Overpass-API#278]
The [wiki says](http://wiki.openstreetmap.org/wiki/Overpass_API/Overpass_QL#newer):

> If the underlying database instance supports attic data, then "changed" is probably a better choice than "newer". 

But,

```
way(318954262);
>;
node._(changed:"2015-04-04T20:42:53Z");
out meta;
```

runs out of memory using the 2Gb allocation on `//overpass-api.de/api/`

On the other hand, "newer" works well and fast:

```
way(318954262);
>;
node._(newer:"2015-04-04T20:42:53Z");
out meta;
```

**Comment by mmd-osm** [comment]
The `(changed: ` variant needs 8.4GB memory to run. It first collects 1 year of changes and only afterwards filters out the result according to the 1549 nodes in the inputset.

**Comment by mmd-osm** [comment]
Out of memory condition could be addressed by the following approach which uses 133MB in total. Instead of collecting all ids, we filter for relevant ids right from the start.

```
<?xml version="1.0" encoding="UTF-8"?>
<osm version="0.6" generator="Overpass API">
<note>The data included in this document is from www.openstreetmap.org. The data is made available under ODbL.</note>
<meta osm_base="2017-09-23T06:13:02Z"/>

After 0h1m3s: in "query", part 5, on line 3. Stack: 0 of 0
  <node id="3569961282" lat="30.7671508" lon="34.8766796" version="1" timestamp="2015-06-03T19:26:13Z" changeset="31701080" uid="2769154" user="shay trails"/>
  <node id="3569961290" lat="30.7674405" lon="34.8766581" version="1" timestamp="2015-06-03T19:26:13Z" changeset="31701080" uid="2769154" user="shay trails"/>
  <node id="4548822059" lat="30.8170173" lon="34.8272000" version="1" timestamp="2016-12-11T08:20:22Z" changeset="44317417" uid="356970" user="valleyofdawn"/>
  <node id="4571723494" lat="30.7715044" lon="34.8325571" version="1" timestamp="2016-12-25T15:31:40Z" changeset="44663543" uid="356970" user="valleyofdawn"/>
  <node id="4833134776" lat="30.6523839" lon="35.0593316" version="1" timestamp="2017-05-03T16:44:02Z" changeset="48370807" uid="356970" user="valleyofdawn"/>
  <node id="4833134781" lat="30.6537470" lon="35.0692333" version="1" timestamp="2017-05-03T16:44:02Z" changeset="48370807" uid="356970" user="valleyofdawn"/>
  <node id="4834421252" lat="30.8185161" lon="34.8029343" version="1" timestamp="2017-05-04T10:19:34Z" changeset="48390172" uid="356970" user="valleyofdawn"/>
  <node id="5034833226" lat="30.5152379" lon="35.0262749" version="1" timestamp="2017-08-14T15:08:25Z" changeset="51113803" uid="356970" user="valleyofdawn"/>
  <node id="5036951263" lat="30.7241734" lon="34.9943417" version="1" timestamp="2017-08-15T13:53:11Z" changeset="51141945" uid="356970" user="valleyofdawn"/>
  <node id="5036963598" lat="30.7768357" lon="34.9272398" version="1" timestamp="2017-08-15T14:02:24Z" changeset="51142166" uid="356970" user="valleyofdawn"/>
  <node id="5039150399" lat="30.7695177" lon="34.9915192" version="1" timestamp="2017-08-16T13:04:50Z" changeset="51171148" uid="356970" user="valleyofdawn"/>
  <node id="5062252071" lat="30.6599129" lon="34.8778663" version="1" timestamp="2017-08-27T14:53:25Z" changeset="51484045" uid="356970" user="valleyofdawn"/>
  <node id="5078471793" lat="30.7628858" lon="34.8769965" version="1" timestamp="2017-09-04T08:44:37Z" changeset="51711430" uid="356970" user="valleyofdawn"/>

</osm>
	Command being timed: "/home/overpass/bla3"
	User time (seconds): 50.26
	System time (seconds): 3.24
	Percent of CPU this job got: 84%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 1:03.00
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 133160
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 32
	Minor (reclaiming a frame) page faults: 34549
	Voluntary context switches: 25928
	Involuntary context switches: 771
	Swaps: 0
	File system inputs: 25199008
	File system outputs: 16
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 0
```




```C++
diff --git a/src/overpass_api/core/datatypes.h b/src/overpass_api/core/datatypes.h
index 30af76e..de887c5 100644
--- a/src/overpass_api/core/datatypes.h
+++ b/src/overpass_api/core/datatypes.h
@@ -474,6 +474,11 @@ struct Change_Entry
     return Id_Type::size_of((uint8*)data + 8) + 8;
   }
 
+  static Id_Type get_id(void* data)
+  {
+    return *(Id_Type*)((uint8*)data + 8);
+  }
+
   void to_data(void* data) const
   {
     old_idx.to_data((uint8*)data);


diff --git a/src/overpass_api/statements/changed.cc b/src/overpass_api/statements/changed.cc
index 59f8a4f..92a0693 100644
--- a/src/overpass_api/statements/changed.cc
+++ b/src/overpass_api/statements/changed.cc
@@ -16,6 +16,7 @@
  * along with Overpass_API.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+#include <functional>
 #include <sstream>
 
 #include "../../template_db/block_backend.h"
@@ -50,6 +51,22 @@ void filter_elems(const std::vector< typename TObject::Id_Type >& ids, std::map<
 }
 
 
+template< class TIndex, class TObject >
+void get_elem_ids(std::vector< typename TObject::Id_Type >& ids, const std::map< TIndex, std::vector< TObject > >& elems)
+{
+  for (typename std::map< TIndex, std::vector< TObject > >::const_iterator it = elems.begin();
+      it != elems.end(); ++it)
+  {
+    std::vector< TObject > local_into;
+    for (typename std::vector< TObject >::const_iterator iit = it->second.begin();
+        iit != it->second.end(); ++iit)
+    {
+        ids.push_back(iit->id);
+    }
+  }
+}
+
+
 template< typename Index, typename Skeleton >
 std::vector< typename Skeleton::Id_Type > collect_changed_elements
     (uint64 since,
@@ -75,6 +92,36 @@ std::vector< typename Skeleton::Id_Type > collect_changed_elements
 }
 
 
+
+template< typename Index, typename Skeleton >
+std::vector< typename Skeleton::Id_Type > collect_changed_elements
+    (uint64 since,
+     uint64 until,
+     Resource_Manager& rman,
+     std::function<bool(typename Skeleton::Id_Type)> pred)
+{
+  std::set< std::pair< Timestamp, Timestamp > > range;
+  range.insert(std::make_pair(Timestamp(since), Timestamp(until)));
+
+  std::vector< typename Skeleton::Id_Type > ids;
+
+  Block_Backend< Timestamp, Change_Entry< typename Skeleton::Id_Type > > changelog_db
+      (rman.get_transaction()->data_index(changelog_file_properties< Skeleton >()));
+  for (typename Block_Backend< Timestamp, Change_Entry< typename Skeleton::Id_Type > >::Range_Iterator
+      it = changelog_db.range_begin(Default_Range_Iterator< Timestamp >(range.begin()),
+            Default_Range_Iterator< Timestamp >(range.end()));
+      !(it == changelog_db.range_end()); ++it)
+    if (pred(it.handle().id()))
+      ids.push_back(it.handle().id());
+
+  std::sort(ids.begin(), ids.end());
+  ids.erase(std::unique(ids.begin(), ids.end()), ids.end());
+  return ids;
+}
+
+
+
+
 //-----------------------------------------------------------------------------
 
 class Changed_Constraint : public Query_Constraint
@@ -168,19 +215,55 @@ void Changed_Constraint::filter(Resource_Manager& rman, Set& into)
 {
   if (!stmt->trivial())
   {
-    std::vector< Node_Skeleton::Id_Type > ids =
-        collect_changed_elements< Uint32_Index, Node_Skeleton >
-        (stmt->get_since(rman), stmt->get_until(rman), rman);
+
+    std::vector< Node_Skeleton::Id_Type > existing_ids;
+    std::vector< Node_Skeleton::Id_Type > ids;
+
+    get_elem_ids(existing_ids, into.nodes);
+    get_elem_ids(existing_ids, into.attic_nodes);
+
+    std::sort(existing_ids.begin(), existing_ids.end());
+    existing_ids.erase(std::unique(existing_ids.begin(), existing_ids.end()), existing_ids.end());
+
+    if (!existing_ids.empty())
+    {
+      ids =
+          collect_changed_elements< Uint32_Index, Node_Skeleton >
+          (stmt->get_since(rman), stmt->get_until(rman), rman,
+              [&] (Node_Skeleton::Id_Type id)
+                {
+                  return (!(id < existing_ids.front() || existing_ids.back() < id) &&
+                           (std::binary_search(existing_ids.begin(), existing_ids.end(), id)));
+                });
+    }
 
     filter_elems(ids, into.nodes);
     filter_elems(ids, into.attic_nodes);
+
   }
 
   if (!stmt->trivial())
   {
-    std::vector< Way_Skeleton::Id_Type > ids =
-        collect_changed_elements< Uint31_Index, Way_Skeleton >
-        (stmt->get_since(rman), stmt->get_until(rman), rman);
+    std::vector< Way_Skeleton::Id_Type > existing_ids;
+    std::vector< Way_Skeleton::Id_Type > ids;
+
+    get_elem_ids(existing_ids, into.ways);
+    get_elem_ids(existing_ids, into.attic_ways);
+
+    std::sort(existing_ids.begin(), existing_ids.end());
+    existing_ids.erase(std::unique(existing_ids.begin(), existing_ids.end()), existing_ids.end());
+
+    if (!existing_ids.empty())
+    {
+      ids =
+          collect_changed_elements< Uint31_Index, Way_Skeleton >
+          (stmt->get_since(rman), stmt->get_until(rman), rman,
+              [&] (Way_Skeleton::Id_Type id)
+              {
+                return (!(id < existing_ids.front() || existing_ids.back() < id) &&
+                         (std::binary_search(existing_ids.begin(), existing_ids.end(), id)));
+              });
+    }
 
     filter_elems(ids, into.ways);
     filter_elems(ids, into.attic_ways);
@@ -188,9 +271,26 @@ void Changed_Constraint::filter(Resource_Manager& rman, Set& into)
 
   if (!stmt->trivial())
   {
-    std::vector< Relation_Skeleton::Id_Type > ids =
-        collect_changed_elements< Uint31_Index, Relation_Skeleton >
-        (stmt->get_since(rman), stmt->get_until(rman), rman);
+    std::vector< Relation_Skeleton::Id_Type > existing_ids;
+    std::vector< Relation_Skeleton::Id_Type > ids;
+
+    get_elem_ids(existing_ids, into.relations);
+    get_elem_ids(existing_ids, into.attic_relations);
+
+    std::sort(existing_ids.begin(), existing_ids.end());
+    existing_ids.erase(std::unique(existing_ids.begin(), existing_ids.end()), existing_ids.end());
+
+    if (!existing_ids.empty())
+    {
+      ids =
+          collect_changed_elements< Uint31_Index, Relation_Skeleton >
+          (stmt->get_since(rman), stmt->get_until(rman), rman,
+              [&] (Relation_Skeleton::Id_Type id)
+              {
+                return (!(id < existing_ids.front() || existing_ids.back() < id) &&
+                         (std::binary_search(existing_ids.begin(), existing_ids.end(), id)));
+              });
+    }
 
     filter_elems(ids, into.relations);
     filter_elems(ids, into.attic_relations);
```

**Comment by drolbr** [comment]
Fixed in 4bc1eeda53e9b0ed74541b63db52a1d9dcc8ce5a.

Nonetheless, the changed command for a long timespan, i.e. more than a month, remains slow. I have adapted the hint in the wiki accordingly.
